### ISSUE TITLE
`INP001` doesn't respect caching behavior

### ISSUE THREAD

**Issue by charliermarsh** [astral-sh/ruff#5449]
See #2075. It looks like the underlying issue (that we don't invalidate the cache when an `__init__.py` is added or removed) is back with the latest cache updates.

_Originally posted by @matthias-roussel-soyhuce in https://github.com/astral-sh/ruff/issues/2075#issuecomment-1612618794_

**Comment by stuthedew** [comment]
Same root cause, reported here for `I001` (isort) rather than `INP001`, since a minimal reproduction is short and both halves are visible in the source.

**Why `I001` is exposed to it.** `match_sources` resolves first-party by probing the *full dotted path* under the `src` roots — `foo.baz` is first-party only if `[SRC]/foo/baz` is a directory or `[SRC]/foo/baz.py` or `.pyi` exists. So one file's `I001` verdict depends on whether a **different** file exists. Meanwhile the per-file cache key is:

```rust
#[derive(CacheKey)]
pub(crate) struct FileCacheKey {
    /// Timestamp when the file was last modified before the (cached) check.
    file_last_modified: FileTime,
    /// Permissions of the file before the (cached) check.
    file_permissions_mode: u32,
}
```

— the linted file's mtime and permission bits, and nothing else, and `Cache::get` compares only a hash of that. Deleting a module therefore invalidates nothing, and every file importing it replays a clean verdict it can no longer earn.

**Reproduction**, identical on ruff 0.16.4 and 0.16.7:

```sh
cd "$(mktemp -d)"
mkdir -p src/pkg/app other
printf '[tool.ruff.lint]\nselect = ["I"]\n' > pyproject.toml
touch src/pkg/__init__.py src/pkg/app/__init__.py
echo "A = 1" > src/pkg/app/gone.py
echo "B = 2" > src/pkg/app/stays.py
printf 'from pkg.app.gone import A\nfrom pkg.app.stays import B\n' > other/probe.py

ruff check .              # All checks passed!
rm src/pkg/app/gone.py
ruff check .              # All checks passed!   <-- stale
ruff check --no-cache .   # I001 on other/probe.py
```

With the cache bypassed the third run reports:

```
I001 [*] Import block is un-sorted or un-formatted
 --> other/probe.py:1:1
```

because `pkg.app.gone` no longer resolves and is sorted third-party while its neighbour stays first-party.

**It is asymmetric, in the direction that hurts.** Deleting a module goes stale; *adding* one is picked up. So the cached answer is wrong precisely when it says "clean" about a tree a fresh checkout rejects — which is how it surfaces in practice: a local `ruff check` passes, CI fails on the same commit, and `rm -rf .ruff_cache` is the only thing that reconciles them.

One wrinkle that makes it easy to misattribute: `touch`ing the *importing* file alone unmasks the diagnostic, with no cache flag, because any edit to it moves the mtime the key reads. We first blamed a stale `__pycache__` for exactly that reason — the experiment that "confirmed" it had also rewritten the importing file.

Happy to split this into its own issue if you would rather keep #5449 to `INP001`.
